File: ios/overlays/HeatMapView.swift
import ExpoModulesCore
import MAMapKit

class HeatMapView: ExpoView {
    var data: [[String: Any]] = []
    var radius: Int = 50
    var opacity: Double = 0.6
    
    private var mapView: MAMapView?
    private var heatmapOverlay: MAHeatMapTileOverlay?
    
    required init(appContext: AppContext? = nil) {
        super.init(appContext: appContext)
    }
    
    /**
     * 设置地图实例
     */
    func setMap(_ map: MAMapView) {
        self.mapView = map
        createOrUpdateHeatMap()
    }
    
    /**
     * 设置热力图数据
     */
    func setData(_ data: [[String: Any]]) {
        self.data = data
        createOrUpdateHeatMap()
    }
    
    /**
     * 设置热力图半径
     */
    func setRadius(_ radius: Int) {
        self.radius = radius
        createOrUpdateHeatMap()
    }
    
    /**
     * 设置透明度
     */
    func setOpacity(_ opacity: Double) {
        self.opacity = opacity
        createOrUpdateHeatMap()
    }
    
    /**
     * 创建或更新热力图
     */
    private func createOrUpdateHeatMap() {
        guard let mapView = mapView else { return }
        
        // 移除旧的热力图
        if let oldHeatmap = heatmapOverlay {
            mapView.remove(oldHeatmap)
        }
        
        // 创建热力图数据
        var heatmapData: [MAHeatMapNode] = []
        for point in data {
            guard let latitude = point["latitude"] as? Double,
                  let longitude = point["longitude"] as? Double else {
                continue
            }
            
            let node = MAHeatMapNode()
            node.coordinate = CLLocationCoordinate2D(latitude: latitude, longitude: longitude)
            node.intensity = 1.0 // 默认强度
            heatmapData.append(node)
        }
        
        guard !heatmapData.isEmpty else { return }
        
        // 创建热力图图层
        let heatmap = MAHeatMapTileOverlay()
        heatmap.data = heatmapData
        heatmap.radius = radius
        heatmap.opacity = CGFloat(opacity)
        
        mapView.add(heatmap)
        heatmapOverlay = heatmap
    }
    
    /**
     * 移除热力图
     */
    func removeHeatMap() {
        guard let mapView = mapView, let heatmap = heatmapOverlay else { return }
        mapView.remove(heatmap)
        heatmapOverlay = nil
    }
    
    override func removeFromSuperview() {
        super.removeFromSuperview()
        removeHeatMap()
    }
}

File: packages/core/android/src/main/java/expo/modules/gaodemap/overlays/HeatMapView.kt
package expo.modules.gaodemap.overlays

import android.content.Context
import android.os.Looper
import android.util.Log
import com.amap.api.maps.AMap
import com.amap.api.maps.model.LatLng
import com.amap.api.maps.model.TileOverlay
import com.amap.api.maps.model.TileOverlayOptions
import com.amap.api.maps.model.HeatmapTileProvider
import com.amap.api.maps.CameraUpdateFactory
import expo.modules.kotlin.AppContext
import expo.modules.kotlin.views.ExpoView
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import expo.modules.gaodemap.utils.LatLngParser

class HeatMapView(context: Context, appContext: AppContext) : ExpoView(context, appContext) {
  
  private val executor: ExecutorService = Executors.newSingleThreadExecutor()
  private val applyUpdateRunnable = Runnable { applyUpdateOnMain() }
  @Volatile private var updateToken: Int = 0
  @Volatile private var needsRebuild: Boolean = true
  private var visible: Boolean = true

  private var heatmapOverlay: TileOverlay? = null
  private var aMap: AMap? = null
  private var dataList: MutableList<LatLng> = mutableListOf()
  private var radius: Int = 50
  private var opacity: Double = 0.6
  
  /**
   * 设置地图实例
   */
  @Suppress("unused")
  fun setMap(map: AMap) {
    aMap = map
    needsRebuild = true
    scheduleUpdate()
  }
  
 

  /**
   * 设置热力图数据
   */
  fun setData(data: List<Any>?) {
    dataList.clear()
    dataList.addAll(LatLngParser.parseLatLngList(data))
    needsRebuild = true
    scheduleUpdate()
  }
  
  /**
   * 设置热力图半径
   */
  fun setRadius(radiusValue: Int) {
    radius = radiusValue
    needsRebuild = true
    scheduleUpdate()
  }
  
  /**
   * 设置透明度
   */
  fun setOpacity(opacityValue: Double) {
    opacity = opacityValue
    applyOverlayOpacity()
  }

  fun setVisible(visibleValue: Boolean) {
    if (!visibleValue) {
      visible = false
      updateToken += 1
      removeCallbacks(applyUpdateRunnable)
      applyOverlayVisibility()
      return
    }

    visible = true
    applyOverlayVisibility()

    if (dataList.isEmpty()) {
      return
    }

    if (heatmapOverlay == null || needsRebuild) {
      scheduleUpdate()
    } else {
      applyOverlayOpacity()
      forceRefresh()
    }
  }

  private fun scheduleUpdate() {
    updateToken += 1
    removeCallbacks(applyUpdateRunnable)
    postDelayed(applyUpdateRunnable, 32)
  }

  private fun applyOverlayVisibility() {
    if (Looper.myLooper() != Looper.getMainLooper()) {
      post { applyOverlayVisibility() }
      return
    }
    val overlay = heatmapOverlay ?: return
    runCatching {
      overlay.javaClass.getMethod("setVisible", Boolean::class.javaPrimitiveType)
        .invoke(overlay, visible)
    }.onFailure {
      if (!visible) {
        overlay.remove()
        heatmapOverlay = null
        needsRebuild = true
      }
    }
  }

  private fun applyOverlayOpacity() {
    if (Looper.myLooper() != Looper.getMainLooper()) {
      post { applyOverlayOpacity() }
      return
    }
    val overlay = heatmapOverlay ?: return
    val opacityValue = opacity.coerceIn(0.0, 1.0)
    val transparency = (1.0 - opacityValue).toFloat()
    runCatching {
      overlay.javaClass.getMethod("setTransparency", Float::class.javaPrimitiveType)
        .invoke(overlay, transparency)
    }
  }

  /**
   * 在主线程应用更新（构建 TileProvider 在后台线程）
   */
  private fun applyUpdateOnMain() {
    if (Looper.myLooper() != Looper.getMainLooper()) {
      post { applyUpdateOnMain() }
      return
    }

    val map = aMap ?: return
    val token = updateToken
    val pointsSnapshot = ArrayList(dataList)
    val radiusValue = radius.coerceIn(10, 200)
    val opacityValue = opacity.coerceIn(0.0, 1.0)

    if (!visible) {
      applyOverlayVisibility()
      return
    }

    if (pointsSnapshot.isEmpty()) {
      heatmapOverlay?.remove()
      heatmapOverlay = null
      return
    }

    if (heatmapOverlay != null && !needsRebuild) {
      applyOverlayVisibility()
      applyOverlayOpacity()
      return
    }

    executor.execute {
      try {
        val provider = HeatmapTileProvider.Builder()
          .data(pointsSnapshot)
          .radius(radiusValue)
          .build()

        post {
          if (token != updateToken) {
            return@post
          }
          if (aMap !== map) {
            return@post
          }
          if (!visible) {
            return@post
          }

          heatmapOverlay?.remove()

          val options = TileOverlayOptions().tileProvider(provider)

          runCatching {
            options.javaClass.getMethod("zIndex", Float::class.javaPrimitiveType)
              .invoke(options, 1f)
          }

          runCatching {
            options.javaClass.getMethod("transparency", Float::class.javaPrimitiveType)
              .invoke(options, (1.0 - opacityValue).toFloat())
          }

          heatmapOverlay = map.addTileOverlay(options)
          needsRebuild = false
          runCatching { heatmapOverlay?.clearTileCache() }
          applyOverlayVisibility()
          applyOverlayOpacity()
          forceRefresh()
        }
      } catch (t: Throwable) {
        Log.e("HeatMapView", "Failed to build heatmap", t)
      }
    }
  }
  
  private fun forceRefresh() {
    runCatching { aMap?.moveCamera(CameraUpdateFactory.zoomBy(0f)) }
  }

  /**
   * 移除热力图
   */
  fun removeHeatMap() {
    updateToken += 1
    removeCallbacks(applyUpdateRunnable)
    heatmapOverlay?.remove()
    heatmapOverlay = null
    dataList.clear()
    needsRebuild = true
  }
  
  override fun onDetachedFromWindow() {
    super.onDetachedFromWindow()
    // 🔑 关键修复：使用 post 延迟检查
    post {
      if (parent == null) {
        removeHeatMap()
        aMap = null
        runCatching { executor.shutdownNow() }
      }
    }
  }
}
